### PR TITLE
Use deepcopy instead of a shallow copy for middlewares, #2349

### PR DESCRIPTION
**Pull request by coder2020official**
#2349

### CODE REVIEW

**Comment by Badiboy** [comment]
I'm not sure that is the best way.

As far as I understand - you want this to allow middleware developer to change values of data in HIS handler. For other cases that is not needed and will just make overhead. 

Is it more reasonable to "say" developers who want to change data values in HIS handler to make deepcopy BEFORE make changes?...

deepcopy is more safe way, but I'm in a humble because I'm not sure that LIBRARY should care about it...

**Comment by coder2020official** [comment]
Well, suppose the dev is adding some data, for post_process.
And this data is not being used in handler.
This will cause the data to be lost. However, if deep copy is used - changes made in the deep copy dictionary in _run_middlewares_and_handlers won't affect original data

File: telebot/async_telebot.py [hunk 1]
```diff
@@ -558,7 +558,7 @@ async def _run_middlewares_and_handlers(self, message, handlers, middlewares, up
                             logger.error("It is not allowed to pass data and values inside data to the handler. Check your handler: {}".format(handler['function']))
                             return
                     else:
-                        data_copy = data.copy()
+                        data_copy = copy.deepcopy(data)
```
Comment (Tishka17) [inline telebot/async_telebot.py:561]: You should no copy data. It is required to pass original objects retrieved from middlewares to handlers

**Comment by coder2020official** [comment]
@Badiboy sorry for the inconvenience, we might need to revert this PR
Could you please do this, since I currently can't

**Comment by coder2020official** [comment]
Or maybe let's wait for me to get back to work, I'll be available to check disadvantages of this again on Sunday
Discard previous comment

**Comment by Badiboy** [comment]
I think it may wait for Sunday.

**Comment by Badiboy** [comment]
As I said from the very beginnig: you are trying to think about all developer's needs yourself. It not looks like a good practice, because in most cases you cannot even imagine all cases how the library is used by all hundreds of thousands of developers.

That's why I always propose as much flexibility for developers as it possible. In this case I still think it's better to leave decision about making deepcopy to developer of middleware...

**Pull request merged**